### PR TITLE
[CVP] Infer nneg on existing zext

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch infers `nneg` flags for existing zext instructions in CVP.
After https://github.com/llvm/llvm-project/pull/71534 and this patch, we can drop `zext -> zext nneg` transform in `RISCVCodeGenPrepare`:

https://github.com/llvm/llvm-project/blob/40671bbdefb6ff83e2685576a3cb041b62f25bbe/llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp#L74-L83

This is an alternative to #72049.

### CODE REVIEW

File: llvm/test/Transforms/CorrelatedValuePropagation/zext.ll [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+for.body:                                         ; preds = %for.cond
+  %ext.wide = zext i32 %a to i64
+  call void @use64(i64 %ext.wide)
+  %ext = trunc i64 %ext.wide to i32
+  br label %for.cond
+
+for.end:                                          ; preds = %for.cond
+  ret void
+}
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/zext.ll:44]: What is the purpose of these loop test cases?

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 2]
```diff
@@ -1032,6 +1033,24 @@ static bool processSExt(SExtInst *SDI, LazyValueInfo *LVI) {
...
 
+static bool processZExt(ZExtInst *SDI, LazyValueInfo *LVI) {
+  if (SDI->getType()->isVectorTy())
+    return false;
+
+  if (SDI->hasNonNeg())
+    return false;
+
+  const Use &Base = SDI->getOperandUse(0);
+  if (!LVI->getConstantRangeAtUse(Base, /*UndefAllowed*/ false)
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1044]: Is there a test case covering at-use?

File: llvm/test/Transforms/CorrelatedValuePropagation/zext.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/CorrelatedValuePropagation/zext.ll:44]: They were copied from @topperc's patch https://github.com/llvm/llvm-project/pull/70715.

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1044]: IIUC the non-loop `@test3` covers at-use.

Comment (nikic) [reply llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1044]: I don't think it does. It has a dominating condition. I would expect something like a select condition for at-use.

File: llvm/test/Transforms/CorrelatedValuePropagation/zext.ll [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+false:
+  br label %exit
+
+exit:
+  %p = phi i32 [ 0, %true.1 ], [ 1, %true.2], [ undef, %false ]
+  %ext = zext i32 %p to i64
+  ret i64 %ext
+}
+
+define i64 @test_infer_at_use(i32 noundef %n) {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/CorrelatedValuePropagation/zext.ll:168]: `noundef` is required here.
Alive2: https://alive2.llvm.org/ce/z/ahASB5
See also https://github.com/llvm/llvm-project/commit/1da0da903ee818385d8ce470d14f9f9c7cdf6942

File: llvm/test/Transforms/CorrelatedValuePropagation/zext.ll [hunk 4]
```diff
@@ -0,0 +1,179 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=correlated-propagation -S | FileCheck %s
+
+; Check that debug locations are preserved. For more info see:
+;   https://llvm.org/docs/SourceLevelDebugging.html#fixing-errors
+; RUN: opt < %s -enable-debugify -passes=correlated-propagation -S 2>&1 | \
+; RUN:   FileCheck %s -check-prefix=DEBUG
+; DEBUG: CheckModuleDebugify: PASS
```
Comment (nikic) [inline llvm/test/Transforms/CorrelatedValuePropagation/zext.ll:8]: Remove these debugify run lines.

File: llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp [hunk 5]
```diff
@@ -1032,6 +1033,24 @@ static bool processSExt(SExtInst *SDI, LazyValueInfo *LVI) {
   return true;
 }
 
+static bool processZExt(ZExtInst *SDI, LazyValueInfo *LVI) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp:1036]:
```suggestion
static bool processZExt(ZExtInst *ZExt, LazyValueInfo *LVI) {
```
Or something? Not sure what SDI is supposed to stand for here.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
Do you plan to implement SCCP support as well?

**Comment by dtcxzyw** [comment]
> Do you plan to implement SCCP support as well?

Yeah. I will post a patch later.

**Pull request merged**